src/Main.ts

```diff
@@ -534,22 +508,84 @@ export class Main {
+                    log.error("Failed to handle DM invite: ", e);
+                    endTimer({outcome: "fail"});
+                }
+            } else if (room) {
+                // Normal invite
+                await room.onMatrixInvite(ev.sender, ev.state_key);
+                endTimer({ outcome: "success" });
+            }
+            return;
+        }
+
+        if (!room) {
+            // We can't do anything else without a room
+            return;
+        }
 
-            try {
-                await this.handleDmInvite(ev.state_key, ev.sender, ev.room_id);
-                endTimer({outcome: "success"});
-            } catch (e) {
-                log.error("Failed to handle DM invite: ", e);
-                endTimer({outcome: "fail"});
+        // Regular membership from matrix user
+        if (!senderIsRemote) {
+            const membership = ev.content.membership;
+            if (membership === "join") {
+                await room.onMatrixJoin(ev.state_key);
+                // Do we need to onboard this user?
+                if (this.config.puppeting?.enabled && this.config.puppeting.onboard_users) {
+                    const adminRoomUser = await this.datastore.getUserAdminRoom(ev.state_key);
+                    const puppets = await this.datastore.getPuppetsByMatrixId(ev.state_key);
+                    if (!adminRoomUser && puppets.length === 0) {
+                        // No admin room, and no puppets but just joined a Slack room.
+                        await UserAdminRoom.inviteAndCreateAdminRoom(ev.state_key, this);
+                    }
+                }
+            } else if (membership === "leave" || membership === "ban") {
+                await room.onMatrixLeave(ev.state_key);
             }
+            // Invites are not handled
+        }
+    }
+
+    public async onMatrixEvent(ev: {
+        event_id: string,
+        state_key?: string,
+        type: string,
+        room_id: string,
+        sender: string,
+        content: any,
+    }) {
+        if (ev.sender === this.botUserId) {
+            // We don't want to handle echo.
             return;
         }
+        // simple de-dup
+        const recents = this.recentMatrixEventIds;
+        for (let i = 0; i < recents.length; i++) {
+            if (recents[i] && recents[i] === ev.event_id) {
+                // move the most recent ev to where we found a dup and add the
+                // duplicate at the end (reasoning: we only want one of the
+                // duplicated ev_id in the list, but we want it at the end)
+                recents[i] = recents[this.mostRecentEventIdIdx];
+                recents[this.mostRecentEventIdIdx] = ev.event_id;
+                log.warn("Ignoring duplicate ev: " + ev.event_id);
+                return;
+            }
+        }
+        this.mostRecentEventIdIdx = (this.mostRecentEventIdIdx + 1) % RECENT_EVENTID_SIZE;
+        recents[this.mostRecentEventIdIdx] = ev.event_id;
```

Comment:
This code is getting hard to understand.
May be time to make a helper class out of this cached list.
It's also hard to understand how this all behaves after a restart and if a busy bridge behaves worse because events are leaving the cache quicker. (I wish we would persist it.)

Reply:
Yeah, I'm wholly unconvinced by this code tbh. I've left it here for now because I'd rather not fight it, but we should use an LRU or something.